### ISSUE TITLE
Implement TF bit-conversion and bit-counting ops

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#211]
Recently there have been several CL's for doing bit counting (count leading zeros, count trailing zeros, population count), and some ops for conversions: today we had RoundUint64ToFloat64().

(My first look at this was that we do not have a single instruction to do this conversion - all our FP conversion instructions, e.g. CVT.D.L, are for signed 64-bit ints. This should be investigated further, and also, maybe we can devise a small macro-asm version that is just a few instructions -- see more below)

Much have these have been specialized to 64-bit arch's (arm64, x64, mips64, ppc64), but there are some related CLs' for 32-bit also. There were comments recently that this is to support 'Wasm'.

Here is a partial list of recent related CL's. Please look at recent commits, and maybe at all TF operators, and see if I have missed any.

```
50d83f90 PPC64: Implemented the RoundUint64ToFloat64 TurboFan operator.
857cd4c1 Implemented the RoundUint64ToFloat64 TurboFan operator for x64 and arm64.
1b535ca0 Implemented the RoundInt64ToFloat32 TurboFan operator for x64, arm64, and mips64. (done)
01fd8e05 Implemented the Word64Popcnt TurboFan operator for x64.
a594ff73 Implemented the ctz Turbo Fan operator for x64.
7558e933 Improved some tests in test-run-machops.
39ed694b Implemented the Word64Clz TurboFan operator for x64, arm64, and mips64. (done)
870e908d [turbofan] Added the RoundInt64ToFloat64 instruction to TurboFan. (done)
053e280c Added Popcnt as an optional operator and implement it on x64 and ia32.
dd1928ca Added a test for CLZ
b3334087 CTZ instruction implemented as optional operator.
```

There are several subtasks here:
- See if there are any that we have instructions for that are now un-implemented, and do them. Add to asm, disasm, sim, etc., as needed. (We should include tests, which I did _not_ do when I implemented 
  dclz instruction for Word64Clz)
- See if any of the _missing_ instructions can be implemented in short sequences, or maybe even longer ones with the out-of-line operations
  - e.g. Uint64ToFloat64, maybe fast-path on values without MSB set (e.g., they fall into range of positive signed int64), and then do slower OOL path for rest.
- Determine if it is a _good_ or _bad_ idea to use multi-instruction sequences.
  - Many of these are optional operators, and the compiler will do the right thing if we don't have the instruction, by emitting longer sequences. Test these to see what code gets generated without the option. We only emit multi-instruction sequences if we can do better!

I'm giving this P1 right now, since there it is a lot of momentum on this in TF, and I don't want to fall behind. We can re-evaluate as we get started..... Assigning to Ákos.

**Comment by paul99** [comment]
There is a brilliant book, called Hacker's Delight by Hank Warren that has all kinds of awesome algoithms & coding tricks. You should buy it if you don't have it. I have hard-copy of first edition (got MIPS overflow detection on add/sub algorithm from that. I have a PDF of 2nd edition, which you can buy here: http://www.informit.com/store/hackers-delight-9780321842688

MIPS has instructions for counting leading zeros and ones, but not for counting trailing zeros and ones, nor for population count.

The book says this will give number of trailing zeros, if you have instruction for number of leading zeros (nlz here, CLZ on MIPS):

```
tailing-zeros = 32 – nlz(~x & (x – 1))
```

There are algorithms for population count also, though looks like a lot of instructions.....

**Comment by paul99** [comment]
Some fix for RoundUint64ToFloat64() has become more urgent, as this was not implemented as an optional operator. So we fail the test on the bots.

**Comment by paul99** [comment]
Ákos did two CL's yesterday:
https://codereview.chromium.org/1439053003/  landed as:
01d77cc3 MIPS64: Implemented the RoundUint64ToFloat64 TurboFan operator for x64 and arm64.
This was landed before the branch was cut, so it is fine.

The second https://codereview.chromium.org/1440133002/ was merged to ahaas' un-landed https://codereview.chromium.org/1438013003/ and will probably go in soon.

So for now, we are done with FP/Int conversion stuff on mips64. 

Please continue with the research on what other bit-counting CL's we should do immediately, and which we need more research on, per the original issue description.

**Comment by kisg** [comment]
We already implemented the remaining bit-conversions in issue #236
